primitives/relayers/src/lane_relayers.rs

```diff
@@ -0,0 +1,625 @@
+			let _ = self.active_set.try_push(next_set.next_set.remove(0));
+		}
+		// finally - remember block where we have activated the set
+		self.enacted_at = current_block;
+
+		true
+	}
+}
+
+/// A set of relayers that will become active at next lane epoch.
+///
+/// The active set of lane relayers is required to change periodically (at `next_set_may_enact_at`).
+/// An interval, when the same relayers set is active is called epoch. Every relayer in the epoch
+/// is guaranteed to have at least one slot, but epochs may have different lengths.
+///
+/// We change the set to guarantee that inactive relayers are removed from the set eventually
+/// and are replaced by active relayers. The relayer will be scheduled for autoremoval if it
+/// has not delivered any messages during previous epoch.
+///
+/// Relayers are bargaining for the place in the set by offering lower reward for delivering
+/// messages. Relayer, which agrees to get a lower reward will likely to replace a "more greedy"
+/// relayer in the `next_set`.
+#[derive(CloneNoBound, Decode, Encode, Eq, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(MaxNextRelayersPerLane))]
+pub struct NextLaneRelayersSet<
+	AccountId: Clone,
+	BlockNumber: Clone,
+	MaxNextRelayersPerLane: Get<u32>,
+> {
+	/// Number of block, where the active set may be replaced with the [`Self::next_set`].
+	///
+	/// We do not allow immediate changes of the active set, because relayers
+	/// may change it so that they are always assigned at the current slot.
+	may_enact_at: BlockNumber,
+	/// Next set of lane relayers.
+	///
+	/// It is a bounded priority queue. Relayers that are working for larger reward are replaced
+	/// with relayers, that are working for smaller reward.
+	next_set: BoundedVec<RelayerAndReward<AccountId>, MaxNextRelayersPerLane>,
+}
+
+impl<AccountId, BlockNumber, MaxNextRelayersPerLane>
+	NextLaneRelayersSet<AccountId, BlockNumber, MaxNextRelayersPerLane>
+where
+	AccountId: Clone + PartialOrd,
+	BlockNumber: Copy,
+	MaxNextRelayersPerLane: Get<u32>,
+{
+	/// Creates new empty relayers set, where next sets enacts at given block.
+	pub fn empty(may_enact_at: BlockNumber) -> Self {
+		NextLaneRelayersSet { may_enact_at, next_set: BoundedVec::new() }
+	}
+
+	/// Returns block, starting from which the `next_set` may be enacted.
+	pub fn may_enact_at(&self) -> BlockNumber {
+		self.may_enact_at
+	}
+
+	/// Set block, starting from which the `next_set` may be enacted.
+	pub fn set_may_enact_at(&mut self, may_enact_at: BlockNumber) {
+		self.may_enact_at = may_enact_at;
+	}
+
+	/// Returns relayer entry from the next set.
+	pub fn relayer(&self, relayer: &AccountId) -> Option<&RelayerAndReward<AccountId>> {
+		self.next_set.iter().find(|r| r.relayer() == relayer)
+	}
+
+	/// Returns relayers from the next set.
+	pub fn relayers(&self) -> &[RelayerAndReward<AccountId>] {
+		self.next_set.as_slice()
+	}
+
+	/// Try insert relayer to the next set.
```

Comment:
Nit: Can we also name the function `try_insert()` ? Since it doesn't act like pushing into a `Vec`. It acts like inserting in a `HashMap`.

Also would be nice to stay consistent with the `HashMap` behavior for [`insert`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.insert) and [`try_insert`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.try_insert). I mean for `try_insert()` to not do anything if there's already an entry for this relayer.

Reply:
> Nit: Can we also name the function try_insert() ? Since it doesn't act like pushing into a Vec. It acts like inserting in a HashMap.

Renamed to `try_insert`.

> Also would be nice to stay consistent with the HashMap behavior for [insert](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.insert) and [try_insert](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.try_insert). I mean for try_insert() to not do anything if there's already an entry for this relayer.

But we need to update value in the set using this function. For example - if relayer has a bid `10`, he may want to change his bid to `5` later. So it behavior looks similar to `insert`. Then why `try_`? Because:
1) it is bounded - so we may not been able to insert if the set is full;
2) it is a priority queue - we may fail to insert if bid is too large.

Reply:
Ok, makes sense.